README.md
# replication-topic-modelling-in-embedding-space

**Ziel des Projekts**
1. Replizieren des Artikels: [Topic Modelling in embedding space](https://arxiv.org/abs/1907.04907) von Adji B. Dieng, Francisco J. R. Ruiz, David M. Blei. Die [originale Implementierung](https://github.com/adjidieng/ETM) von Autoren. 
2. Neues Experiment: Kombination von pre-fitted BERT-Embeddings mit ETM
**Bericht**
1. Berichtschreiben im Overleaf: [Bericht](https://www.overleaf.com/read/wpfpwbxwwjhz), [HTML-Notebook-as-pdf](https://github.com/hanhluukim/replication-topic-modelling-in-embedding-space/blob/main/pdf_html_files/notebook_replication.pdf)

**Durchgeführte Experimenten**
- [x] Laufzeit des ETM Modells in der Abhängigkeit von der Vokabulargrößen
- [x] Vergleich zwischen LDA und ETM auf dem Datensatz: 20NewsGroups (Maß: Topic Quality = Topic Coherence * Topic Diversity, normalisierte-Perplexity)
- [x] Vergleich Ergebnisse von ETM auf 20NewsGroups mit Stopwords und ohne-Stopwords
- [x] Vergleich der Embedding zwischen Skipgramm und BERT-Wortembedding
- [x] Vergleich zwischen prefitted-Skipramm-ETM und prefitted-BERT-ETM

**Verwandte Pakette**
1. List aller verwendeten Paketten ist in der Datei: requirements.txt
2. Installieren durch: `pip install -r requirements.txt`
3. Wichtig sind: [gensim](https://radimrehurek.com/gensim/) für LDA, [torch](https://pytorch.org/docs/stable/torch.html) für variationale Inference, [sklearn](https://scikit-learn.org/stable/) für Datensatz
4. Anmerkung: Unsere Implementierung verwendet gensim=="3.8.3", sodass einige Parameters andere Namen hat als in dem neusten gensim (gensim=="4.2.0")

**Notwendige Dateien für BERT-prefitted-ETM**
1. Herunterladen die Bert-Embeddings from [Google-Drive-Link](https://drive.google.com/file/d/1aLLQCDFncdaedOS4pnB0-T6y0T7dHIck/view?usp=sharing): 
2. Packen die txt.Datei in dem Ordner: prepared_data/
3. Durchführen in dem Projektsordner: `python src/bert_covert_format.py`. Zwei Dateein: bert_embeddings.npy und bert_vocab.txt wurden automatisch in dem Ordner: `prepared_data/` gespeichert. 
4. Diese beiden Dateien sind notwendig für das Bert-prefitted-ETM

**Befehlen zum Durchführung von Experimenten**
1. LDA-Modelle: `python main_lda.py --filter-stopwords "True" --min-df 100 --epochs 5 --use-tensor True --batch-test-size 100` 
2. ETM-Modelle: `python main.py --model "ETM" --epochs 160 --wordvec-model "bert" --loss-name "cross-entropy" --min-df 2 --num-topics 20 --filter-stopwords "True" --hidden-size 800 --activate-func "ReLU" --optimizer-name "adam" --lr 0.003 --wdecay 0.0000012`
3. `--wordvec-model` kann folgende Werten haben: "cbow", "skipgram", "bert", `--optimizer-name` hat folgende Optionen "adam" und "sgd"

**Notebooks zum Durchführung von Experimenten**
1. [Notebook für LDA](https://github.com/hanhluukim/replication-topic-modelling-in-embedding-space/blob/main/notebook_topic_modelling_with_LDA.ipynb)
2. [Notebook für ETM](https://github.com/hanhluukim/replication-topic-modelling-in-embedding-space/blob/main/notebook_replication.ipynb)
3. [Notebook für Wordembeddings-BERT](https://github.com/hanhluukim/replication-topic-modelling-in-embedding-space/blob/main/notebook_bert_sentence_embeddings_to_word_embeddings.ipynb)
4. [Notebook für semantische Ähnlichkeiten](https://github.com/hanhluukim/replication-topic-modelling-in-embedding-space/blob/main/notebook_comparison_embeding_models.ipynb)

**Struktur des Repos**
1. Implementierung befindet sich im Ordner [src](https://github.com/hanhluukim/replication-topic-modelling-in-embedding-space/tree/main/src)
2. Jeder Schritt von den durchgeführten Experimenten kann in der Datei [notebook_replication.ipynb](https://github.com/hanhluukim/replication-topic-modelling-in-embedding-space/blob/main/notebook_replication.ipynb) nachvollzieht und kontrolliert werden. Man kann dieses Notebook selbst auf dem Link Colab öffnen und durchführen. 
3. Jeder Schritt für Erstellung von Word-embeddings mit BERT in der Datei [notebook_bert_embedding.jpynb](https://github.com/hanhluukim/replication-topic-modelling-in-embedding-space/blob/main/notebook_bert_sentence_embeddings_to_word_embeddings.ipynb)

**Genutzte Datensätze**
1. [20NewsGroups]()
2. Vorverarbeitungsschritten für Raw-Datensatz: Entfernen von nicht-alphabeltischen Zeichen, Stoppwörtern, Kleinschreiben, ...
3. Zerlegung: Traindatensatz (112114), Testdatensatz (Testdatensatz-h1, Testdatensatz-h2) (7532) und Validationssatz von 100 Dokumenten

**Angewandte Word-Embedding Methoden**
- [x] CBOW und Skipgram mittels [Gensim](https://radimrehurek.com/gensim/). Word-Embeddings werden für jedes Wort des Traindatensatzes gelernt. Es wird ein Trainingsprozess gebraucht, um Word-Repräsentationen zu haben
- [x] based-BERT-Wordembedding mittels [Transformer-Huggingface](https://huggingface.co/docs/transformers/installation). Benutzung von pretrained-BERT-Modell um direkt Repräsentationen für Wörter zu haben. Kein Trainieren ist benötig. 
- [x] Implementierung in : `src/embedding.py` für CBOW und Skipgram und für BERT: `src/bert_preparing.py`, `src/bert_embedding.py`, `bert_main.py`. Durchfühen Bert-Embedding mit `python bert_main.py` aus dem Hauptordner

**Architektur des ETM-Modells**
1. Einige wichtige Beschreibung:

- Das Vocabular V: besteht aus den einzigartigen Wörtern aus dem Traindatensatz
- Eingaben für ETM: (1) normalisierte-BOW für Encoder-Network und (2) Vocabular-Word-Embeddings
- Topic Embedding ist numersiche Repränstation für ein Topic in dem Embedding-Space, in dem Wort-Embeddings sich auch befinden

2. Das Modell ETM wird in `src/etm.py` implementiert.


history.txt
* b15fc85 (HEAD -> main, origin/main, origin/HEAD) (update completed notebook for using, 2022-05-30)
* 6643fd8 (update READme, 2022-05-30)
* 9759c68 (create colab notebook for control preparing data, 2022-05-29)
* 6efaa03 (get_perplexity problem, 2022-05-29)
* 23e02a5 (update to device get_perplexity, 2022-05-29)
* a898d4b (update to device get_perplexity, 2022-05-29)
* 197e834 (update to device get_perplexity, 2022-05-29)
* cfd0012 (update to device get_perplexity, 2022-05-29)
* f9fbc6b (add to(device) to perplexity, 2022-05-29)
* f705502 (clean folders update visualization, 2022-05-29)
* 134370d (run time figure, 2022-05-29)
* 216388b (update main_lda.py show ppl each batch, 2022-05-29)
* 8e0372b (remove etm-julia from dir, 2022-05-29)
* c054859 (remove testing_notebooks, 2022-05-29)
* 246e4a7 (remove testing_notebooks, 2022-05-29)
* 1953962 (Update README.md, 2022-05-29)
* 9bb13d8 (Update README.md, 2022-05-29)
* 8a8092a (Update README.md, 2022-05-29)
* 566c9a7 (Update README.md, 2022-05-29)
* bcb00ee (perplexity update, update notebook, 2022-05-29)
* 0a463e2 (update perplexity for lda, main_lda.py command, add new topicPerplexityNew, 2022-05-28)
* 4397f69 (Update README.md, 2022-05-28)
* f6da1e4 (Update README.md, 2022-05-28)
* 1e7cb84 (add lda-perplexity as batches of 100, 2022-05-28)
* 266481b (visualization runtime update path figures, 2022-05-28)
* d113463 (visualization runtime update path figures, 2022-05-28)
* 725156b (visualization runtime, update normalized perplexity for ETM, 2022-05-28)
* cfab556 (bert problem solved, update not_in_bert_vocab, update read_prefitted for bert-embedding, 2022-05-28)
* d202402 (perplexity build in, 2022-05-28)
* d136413 (solved bert-vocab for all min_df, 2022-05-27)
*   83b0861 (Merge branch 'main' of https://github.com/hanhluukim/replication-topic-modelling-in-embedding-space into main, 2022-05-27)
|\  
| * 7255c3f (notebook for complete running from data to etm, 2022-05-27)
* | f3c36f3 (tiny change self.train(), 2022-05-27)
|/  
* c2a1a5e (testdataset check, 2022-05-27)
* 3ddf364 (add visualization=True to train(), 2022-05-27)
* fb3e5f6 (run notebook main.py with all args, 2022-05-27)
* f05ded7 (control args.lr, args.wdecay, update plot for neg-kld, 2022-05-27)
* ad485e3 (change KLD plot, add args.lr, args.wdecay to main, 2022-05-27)
* d6428dc (run bert-prefitted-ETM, save figures, topics by word2vec_model, add comments to main, 2022-05-27)
* 710fcff (update notebook etm, 2022-05-27)
* 52f5d32 (solved conflict vocab file, 2022-05-27)
* 3373c58 (ETM error found at cross-entropy, 2022-05-27)
* 1dbc897 (remove sorted train_indices from get_doc_in_words, 2022-05-26)
* 7670261 (control prepared_data, update read_prefitted, todo testdataset, 2022-05-27)
* 479d205 (update notebook main and notebook embeddings, 2022-05-26)
* 5fa13e5 (update notebook_replication cause of implementation changes in /src, 2022-05-26)
* 2f64a47 (results etm by hidden-size and activat func, 2022-05-26)
* 4ba5530 (add new args to main for hidden-size and activat func, 2022-05-26)
* 0051ac3 (add new args to main, save etm-topics with topics und without topics directory, 2022-05-26)
* f0fc8a9 (add new args to main args.filter_stopwords and use_bert_embedding, 2022-05-26)
* 459043a (lda new runs stopwords and without stopwords, 2022-05-26)
* c6ef450 (add not_in_bert_vocab, update prepare_dataset.py, 2022-05-26)
* 38d787a (lda run stopwords and without stopwords, 2022-05-25)
*   08a59a4 (update epochs for LDA, 2022-05-25)
|\  
| *   17ed1e8 (das was schon oben war übernehmen, 2022-05-25)
| |\  
| * | fbf97cf (Fehlerbehebung in Perplexity, 2022-05-25)
* | | d3798bf (update epochs for LDA, 2022-05-25)
| |/  
|/|   
* | 5be92eb (move runtime files, 2022-05-25)
* | e3bb55a (notebook compare similar words of different embeddings methods, 2022-05-25)
* | c41c744 (remove old BERT notebook, remove old lda corpus function, 2022-05-25)
* | 4d74944 (update documentation for implementation words-embeddings with BERT, 2022-05-25)
* | 742b689 (control of correctness of implementation with bert, 2022-05-25)
* | b0d7dfa (add not_in_bert_vocab, 2022-05-25)
* | a036309 (LDA results by min_df, notebooks documentation, 2022-05-25)
|/  
* b01b7d6 (fixing lda gensim error, 2022-05-25)
* 84a9103 (etm and lda num_topis 20 min_df 100, 2022-05-25)
* ac822d2 (etm and lda num_topis 20 min_df 100, 2022-05-25)
* 6316f8b (etm and lda num_topis 20, 2022-05-25)
* a13da3b (etm min_df 2 num_topis 50, 2022-05-24)
* 27fe96b (etm min_df 2 10, 20 epochs, 2022-05-24)
* 5886160 (etm 200 n-topcs 10 epochs results, 2022-05-24)
* be46c13 (etm 100 epochs results, 2022-05-24)
* 5153326 (add torch to embedding.py, 2022-05-24)
* f6363ed (update main_eval_checkpoints, update seed=42 embeddings, 2022-05-24)
* b07dc3a (change args.min_df, 2022-05-24)
* 0faf952 (epochs 100, update main.py with num_toics, update train_etm save by num_topics, add load ckpt, 2022-05-24)
* 981674b (remove textsloader from main_lda, 2022-05-24)
* c9395c7 (add loss, min_df args to main.py, update read_prefitted, update find similar words, 2022-05-24)
* b9f0386 (update bert_preparing remove stopwords list, 2022-05-24)
* c81ab18 (update bert preparing fix splitting long sentences, 2022-05-24)
* 12dfa7b (solved similar problem, 2022-05-24)
* 501dab7 (remove cache topics, 2022-05-24)
* acc32b3 (update gitignore file, 2022-05-24)
* 812765e (problem lda by min_df, 2022-05-24)
* cc4471b (problem mit lda, update evaluierung with number words, 2022-05-24)
* 3e0a268 (not upload min_df dirs, 2022-05-23)
* 91b0010 (remove saving, update ignore file, run notebook, 2022-05-23)
* 1a0cc48 (testing save_path in notebook, 2022-05-23)
* 0ecd32d (embedding.py change save path, 2022-05-23)
* 897a45a (remove save prepared_data/ outputs from code, 2022-05-23)
* fa436be (add checkpoints, 2022-05-23)
* 5bc6be4 (push etm_julia, 2022-05-23)
* 9513197 (test gitignore after rm --cached, 2022-05-23)
* cfb5892 (add prepared_data/ in gitignore, 2022-05-23)

main.py
from src.prepare_dataset import TextDataLoader
import argparse
from collections import Counter
import pandas as pd
from src.embedding import WordEmbeddingCreator, read_prefitted_embedding
from pathlib import Path
from src.train_etm import DocSet, TrainETM
from src.etm import ETM
import torch
from datetime import datetime
#from src.visualization import show_embedding_with_kmeans_umap
import subprocess
import numpy as np
import random
from src.evaluierung import topicCoherence2, topicDiversity
from tqdm import tqdm
import math
from torch.utils.data import DataLoader
from src.utils_perplexity import get_perplexity


#--------------------deterministic------------------------------------
import os
seed = 42
os.environ['PYTHONHASHSEED'] = str(seed)
torch.manual_seed(seed)
torch.cuda.manual_seed(seed)
torch.cuda.manual_seed_all(seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
np.random.seed(seed)
random.seed(seed)



#---------------------check cuda-------------------------------------
device = torch.device("cuda" if torch.cuda.is_available() else "cpu") 
print(f'using cuda: {torch.cuda.is_available()}')

#----------------------argument--------------------------------------
parser = argparse.ArgumentParser(description='main.py')
parser.add_argument('--model', type=str, default="LDA", help='which topic model should be used')
parser.add_argument('--epochs', type=int, default=150, help='train epochs')
parser.add_argument('--wordvec-model', type=str, default="skipgram", help='method for training word embedding')
parser.add_argument('--loss-name', type=str, default="cross-entropy", help='loss-name')
parser.add_argument('--min-df', type=int, default=10, help='minimal document frequency for vocab size')
parser.add_argument('--num-topics', type=int, default=10, help='number of topics')
parser.add_argument('--filter-stopwords', type=str, default="True", help='filter or not filter stopwords')
parser.add_argument('--hidden-size', type=int, default=100, help='linear transformation')
parser.add_argument('--activate-func', type=str, default="tanh", help='tanh oder ReLU activate function')
parser.add_argument('--optimizer-name', type=str, default="adam", help='adam or sgd')
parser.add_argument('--lr', type=float, default=0.002, help='learning rate')
parser.add_argument('--wdecay', type=float, default=0.0000012, help='weights decay')

args = parser.parse_args()


#------------------------parser--------------------------------------
model_name = args.model
epochs = args.epochs
word2vec_model = args.wordvec_model
if word2vec_model=="bert":
  use_bert_embedding = True
else:
  use_bert_embedding = False

min_df = args.min_df
num_topics = args.num_topics
n_hidden_size = args.hidden_size
activate_func = args.activate_func
optimizer_name = args.optimizer_name
learning_rate = args.lr
wdecay = args.wdecay

filter_stopwords = args.filter_stopwords
if filter_stopwords == "True":
  topics_under_dir = "no_stopwords"
  stopwords_filter = True
else:
  topics_under_dir = "with_stopwords"
  stopwords_filter = False
print(f'filter-stopwords: {stopwords_filter}')


#-----------------------check statistic--------------------------------
"""
for min_df in [2, 5, 10, 30, 100]:
      Path('prepared_data').mkdir(parents=True, exist_ok=True)
      Path(f'prepared_data/min_df_{min_df}').mkdir(parents=True, exist_ok=True)
      
      textsloader = TextDataLoader(source="20newsgroups", train_size=None, test_size=None)
      textsloader.load_tokenize_texts("20newsgroups")
      textsloader.preprocess_texts(length_one_remove=True, punctuation_lower = True, stopwords_filter = True)
      textsloader.split_and_create_voca_from_trainset(max_df=0.7, min_df=min_df, stopwords_remove_from_voca=True)
      for_lda_model = False
      word2id, id2word, train_set, test_set, val_set = textsloader.create_bow_and_savebow_for_each_set(for_lda_model=for_lda_model, normalize = True)
      textsloader.write_info_vocab_to_text()
      del textsloader 
""" 
#------------------------prepare data---------------------------------
print(100*"-")
textsloader = TextDataLoader(source="20newsgroups", train_size=None, test_size=None)
print("\n")
textsloader.load_tokenize_texts("20newsgroups")

#-------------------------preprocessing-------------------------------

textsloader.preprocess_texts(length_one_remove=True, 
                             punctuation_lower = True, 
                             stopwords_filter = stopwords_filter,
                             use_bert_embedding = use_bert_embedding)
#textsloader.show_example_raw_texts(n_docs=2)
print("\ntotal documents {}".format(len(textsloader.complete_docs)))

textsloader.split_and_create_voca_from_trainset(max_df=0.7, 
                                                min_df=min_df, 
                                                stopwords_remove_from_voca= stopwords_filter)
print("\n")

#-------------------------------test data for ETM--------------------------------------
for_lda_model = False
word2id, id2word, train_set, test_set, val_set = textsloader.create_bow_and_savebow_for_each_set(for_lda_model=for_lda_model, 
                                                                                                 normalize = True)
print("train-bow-representation for ETM: \n")
print(f'example ids of dict-id2word for ETM: {list(id2word.keys())[:5]}')
print(f'example words of dict-id2word for ETM: {list(id2word.values())[:5]}')
textsloader.write_info_vocab_to_text()


#------------------------------short summary information-------------------------------
print(f'Size of the vocabulary after prprocessing ist: {len(textsloader.vocabulary)}')
print(f'Size of train set: {len(train_set["tokens"])}')
print(f'Size of val set: {len(val_set["tokens"])}')
print(f'Size of test set: {len(test_set["test"]["tokens"])}')

#del test_set
del val_set

#------------------------get docs in words to use for training embedding---------------
#------------------------doc in words for embedding training---------------------------
#------------------------re-erstellen von Dokumenten nach der Vorverarbeitungen.-------
#------------------------Die Dokumenten sind in Wörtern und werden für Word-Embedding Training benutzt

docs_tr, docs_t, docs_v = textsloader.get_docs_in_words_for_each_set()
del textsloader
print('prepare data finished')
print(100*"-")

#------------------------paths-------------------------------------------------
print('word-embedding training begin')
save_path = Path.joinpath(Path.cwd(), f'prepared_data/min_df_{min_df}')
if word2vec_model == "bert":
  figures_path = Path.joinpath(Path.cwd(), f'figures/bert_min_df_{min_df}')
else:
  figures_path = Path.joinpath(Path.cwd(), f'figures/min_df_{min_df}')

Path(save_path).mkdir(parents=True, exist_ok=True)
Path(figures_path).mkdir(parents=True, exist_ok=True)

vocab = list(word2id.keys())
#-------------------------embedding training------------------------------------------
if word2vec_model!="bert" and use_bert_embedding == False:
      wb_creator = WordEmbeddingCreator(model_name=word2vec_model, documents = docs_tr, save_path= save_path)
      wb_creator.train(min_count=0, embedding_size= 300)
      wb_creator.create_and_save_vocab_embedding(vocab, save_path)
      #wb_creator.cluster_words(save_path, figures_path , 2)
      # show embedding of some words
      print("neighbor words of some sample selected words")
      print(f'word: {vocab[0]}')
      print(f'vector: {list(wb_creator.model.wv.__getitem__(vocab[0]))[:5]} ')
      print(wb_creator.find_most_similar_words(n_neighbor=5, 
                                               word=vocab[0]))

else:
      #todo run subprocess
      print("using prepared_data/bert_vocab_embedding.txt")
      print("bert-embeddings were already builded")
      """
      subprocess.run(
            ["python", "bert_main.py"])
      """     
print("word-embedding finised")
print(100*"-")
#--------------------------topic embedding training-----------------------------------

#embedding word-vectors visualize
#embedding_path = save_path
#fig_path = Path.joinpath(Path.cwd(), "figures")
#wb_creator.cluster_words(embedding_path, fig_path, 2)

print('training parameter setting...')
# setting parameters for training ETM
class TrainArguments:
      def __init__(self, epochs, batch_size, log_interval):
          self.epochs = epochs
          self.batch_size = batch_size
          self.log_interval = log_interval
class OptimizerArguments:
      def __init__(self, optimizer_name, lr, wdecay):
            self.optimizer = optimizer_name
            self.lr = lr
            self.wdecay = wdecay

train_args = TrainArguments(epochs=epochs, 
                            batch_size=1000, 
                            log_interval=None)
optimizer_args = OptimizerArguments(optimizer_name= optimizer_name, 
                                    lr=learning_rate, 
                                    wdecay=wdecay)

print(f'using epochs: {train_args.epochs}')
print(f'using optimizer: {optimizer_args.optimizer}')
print(f'using learning rate: {optimizer_args.lr}')
print(f'using wdecay: {optimizer_args.wdecay}')

#--------------------------using Dataset Modul to create DocSet-------------------------

vocab_size = len(list(word2id.keys()))
tr_set = DocSet("train", vocab_size, train_set, normalize_data=True)
print(f'total train docs: {len(tr_set)}')
print(f'sum of vector: {sum(tr_set.__getitem__(0)["normalized"])}')
print(f'length of vector: {torch.norm(tr_set.__getitem__(0)["normalized"])}')


#---------------------------reading embedding data from file----------------------------
print(f'reading {word2vec_model} prefitted-embedding...')
if word2vec_model == "bert" and use_bert_embedding == True:
  embedding_vocab, embedding_data = read_prefitted_embedding(word2vec_model, vocab, "prepared_data/")
else:
  embedding_vocab, embedding_data = read_prefitted_embedding(word2vec_model, vocab, save_path)

w = vocab[0]
idx = embedding_vocab.index(w)
print(f'example 5 element of word-vector: {embedding_data[idx][:5]}')
del embedding_vocab

#---------------------------etm-model setting parameters--------------------------------
#num_topics = 50
t_hidden_size = n_hidden_size #args.hidden_size
rho_size = len(embedding_data[0])
emb_size = len(embedding_data[0])
theta_act = activate_func #args.activate_func #"ReLU"

#--------------------------etm model setting-------------------------------------------
print("ETM initilize...")
etm_model = ETM(
  num_topics, vocab_size, t_hidden_size, rho_size, emb_size, theta_act, 
  embedding_data, enc_drop=0.5).to(device)
print(50*"-" + 'MODEL-SUMMARY' + 50*"-")
print(etm_model)


# train_set must be normalized??
"""
def get_normalized_bows(dataset):
    
    return dataset
    
train_set = get_normalized_bows(train_sprint("end train time: {}".format(datetime.now()-start_time))et)
#
"""

#--------------------------training----------------------------------------------------
print(50*"-" + 'TRAIN' + 50*"-")

start = datetime.now()
train_class = TrainETM().train(
    etm_model,
    args.loss_name,
    vocab_size, 
    train_args, optimizer_args, train_set,
    normalize_data = True,
    figures_path = figures_path,
    num_topics = num_topics,
    visualization=True)
    

#--------------------------------RUN TIME------------------------------------------------
f = open('info_run_time/python_train_runtime.txt', 'a')
f.write(f'train with prefitted-{word2vec_model} min_df: {min_df} \t vocab-size {len(vocab)} \t epochs: {epochs} \t run time: {datetime.now()-start}\n')
f.close()

#-----------------------------------show topics---------------------------------
etm_model.eval()
topics = etm_model.show_topics(id2word, 25)
topics = [[e[0] for e in tp] for tp in topics] #get only top words

#-------------------------------save topics and evaluation-----------------------
if word2vec_model == "bert":
  save_topics_path = f'topics/{topics_under_dir}/min_df_{min_df}/bert_etm'
else:
  save_topics_path = f'topics/{topics_under_dir}/min_df_{min_df}/etm'
#-------------------------------------------------------------------------------
Path(save_topics_path).mkdir(parents=True, exist_ok=True)
topics_f = open(f'{save_topics_path}/{num_topics}_topics.txt', 'w')
topic_idx = 0
for tp in tqdm(topics): 
    tp_as_str = " ".join(tp[:10])
    topics_f.write(f'topic: {topic_idx + 1} {tp_as_str}\n')
    topic_idx += 1
topics_f.close()

# topic coherence and topic diversity and quality
tc = topicCoherence2(topics,len(topics),docs_tr,len(docs_tr))
td = topicDiversity(topics)
print(f'topic-coherrence: {tc}')
print(f'topic-diversity: {td}')


_, e_ppl = get_perplexity(etm_model, test_set, vocab_size, 1000)
print(f'topic-normalized-perplexity: {e_ppl}')

eval_f = open(f'{save_topics_path}/{num_topics}_evaluation.txt', 'a')
eval_f.write(f'epochs \t n-hidden-size \t activate-func \t optimizer \t coherrence \t diversity \t quality \t perplexity\n')
eval_f.write(f'{epochs} \t{t_hidden_size} \t {activate_func} \t {optimizer_name} \t {tc} \t {td} \t {tc*td} \t {e_ppl}\n')
eval_f.close()

del etm_model
del train_class 

#----------------
"""
show_embedding_with_kmeans_umap(
  id2word, embedding_data, num_topics, etm_model.topic_embeddings_alphas.weight,
  figures_path)
"""